File: packages/KNIRVSDK/ts/unified/src/wallet.ts
/** Controller-custodied wallet client. Private user keys never enter the SDK process. */
import type { DirectSignRequest, KNIRVFee } from './signing';

export interface WalletResponse<T = unknown> {
  code: number;
  status: WalletResponseStatus;
  type: WalletResponseType;
  message?: string;
  data?: T;
}

export enum WalletResponseStatus { SUCCESS = 'success', FAILURE = 'failure', REJECT = 'reject' }
export enum WalletResponseType { ESTABLISH = 'establish', ACCOUNT = 'account', NETWORK = 'network', SIGN = 'sign', TRANSACTION = 'transaction' }
export enum WalletResponseExecuteType { ADD_ESTABLISH = 'ADD_ESTABLISH', GET_ACCOUNT = 'GET_ACCOUNT', ADD_NETWORK = 'ADD_NETWORK', SWITCH_NETWORK = 'SWITCH_NETWORK', DO_CONTRACT = 'DO_CONTRACT', SIGN_TX = 'SIGN_TX' }
export enum WalletResponseFailureType { NETWORK_TIMEOUT = 'NETWORK_TIMEOUT', UNAPPROVED_CHAIN = 'UNAPPROVED_CHAIN', UNAPPROVED_HOST = 'UNAPPROVED_HOST', LOCKED_ACCOUNT = 'LOCKED_ACCOUNT', INVALID_FORMAT = 'INVALID_FORMAT', INVALID_TRANSACTION = 'INVALID_TRANSACTION', UNEXPECTED_ERROR = 'UNEXPECTED_ERROR' }
export enum WalletResponseRejectType { ESTABLISH_REJECTED = 'ESTABLISH_REJECTED', SIGN_REJECTED = 'SIGN_REJECTED', TRANSACTION_REJECTED = 'TRANSACTION_REJECTED' }
export enum WalletResponseSuccessType { ESTABLISH_SUCCESS = 'ESTABLISH_SUCCESS', SIGN_SUCCESS = 'SIGN_SUCCESS', TRANSACTION_SUCCESS = 'TRANSACTION_SUCCESS' }

export interface SkillInvocationParams {
  user_id: string;
  skill_id: string;
  sender: string;
  amount: string;
  account_number: string | number | bigint;
  sequence: string | number | bigint;
  fee?: KNIRVFee;
  metadata?: Record<string, unknown>;
}

export interface KNIRVWalletConfig {
  provider?: string;
  chainId?: string;
  enableGasless?: boolean;
  apiKey?: string;
  approvalTimeoutMs?: number;
}

const DEFAULT_GATEWAYS = [
  'https://gateway.knirv.network',
  'https://testnet-gateway.knirv.network',
  'http://localhost:8080',
];

interface ApprovalRequest {
  request_id: string;
  approval_uri: string;
  expires_at: string;
}

interface ApprovalStatus {
  status: 'pending' | 'approved' | 'rejected' | 'expired';
  result?: unknown;
  reason?: string;
}

function approvalJSON(value: unknown): string {
  return JSON.stringify(value, (_key, item) => {
    if (typeof item === 'bigint') return item.toString();
    if (!(item instanceof Uint8Array)) return item;
    let binary = '';
    for (let start = 0; start < item.length; start += 0x8000) {
      binary += String.fromCharCode(...item.subarray(start, start + 0x8000));
    }
    return btoa(binary);
  });
}

export class KNIRVWallet {
  private readonly gateways: string[];
  private readonly chainId: string;
  private readonly apiKey?: string;
  private readonly approvalTimeoutMs: number;

  constructor(config: KNIRVWalletConfig = {}) {
    this.gateways = config.provider
      ? [config.provider.replace(/\/$/, ''), ...DEFAULT_GATEWAYS.filter((url) => url !== config.provider)]
      : DEFAULT_GATEWAYS;
    this.chainId = config.chainId || 'knirv-1';
    this.apiKey = config.apiKey;
    this.approvalTimeoutMs = config.approvalTimeoutMs || 5 * 60_000;
  }

  private async request<T>(path: string, init: RequestInit = {}): Promise<T> {
    let lastError: unknown;
    for (const gateway of this.gateways) {
      try {
        const response = await fetch(`${gateway}${path}`, {
          ...init,
          headers: {
            'Content-Type': 'application/json',
            ...(this.apiKey ? { Authorization: `Bearer ${this.apiKey}` } : {}),
            ...init.headers,
          },
        });
        if (response.status >= 500) throw new Error(`${gateway} returned HTTP ${response.status}`);
        if (!response.ok) throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        return await response.json() as T;
      } catch (error) {
        lastError = error;
      }
    }
    throw lastError instanceof Error ? lastError : new Error('No KNIRV gateway is available');
  }

  private async approvedOperation(path: string, payload: unknown, type: WalletResponseType): Promise<WalletResponse> {
    const approval = await this.request<ApprovalRequest>(path, { method: 'POST', body: approvalJSON(payload) });
    const deadline = Date.now() + this.approvalTimeoutMs;
    while (Date.now() < deadline) {
      const state = await this.request<ApprovalStatus>(`/api/controller/signing/requests/${encodeURIComponent(approval.request_id)}`);
      if (state.status === 'approved') return { code: 0, status: WalletResponseStatus.SUCCESS, type, data: state.result };
      if (state.status === 'rejected') return { code: 1, status: WalletResponseStatus.REJECT, type, message: state.reason || 'Rejected in KNIRVCONTROLLER' };
      if (state.status === 'expired') return { code: 1, status: WalletResponseStatus.FAILURE, type, message: 'KNIRVCONTROLLER approval expired' };
      await new Promise((resolve) => setTimeout(resolve, 1500));
    }
    return { code: 1, status: WalletResponseStatus.FAILURE, type, message: 'KNIRVCONTROLLER approval timed out' };
  }

  async getAccount(): Promise<WalletResponse> {
    const data = await this.request('/api/controller/account');
    return { code: 0, status: WalletResponseStatus.SUCCESS, type: WalletResponseType.ACCOUNT, data };
  }

  async addNetwork(params: unknown): Promise<WalletResponse> {
    return this.approvedOperation('/api/controller/networks', params, WalletResponseType.NETWORK);
  }

  async switchNetwork(chainId: string): Promise<WalletResponse> {
    return this.approvedOperation('/api/controller/networks/switch', { chain_id: chainId }, WalletResponseType.NETWORK);
  }

  async signTransaction(transaction: DirectSignRequest): Promise<WalletResponse> {
    return this.approvedOperation('/api/controller/signing/requests', { kind: 'transaction', chain_id: this.chainId, transaction }, WalletResponseType.SIGN);
  }

  async signMessage(envelope: unknown): Promise<WalletResponse> {
    return this.approvedOperation('/api/controller/signing/requests', { kind: 'message', chain_id: this.chainId, envelope }, WalletResponseType.SIGN);
  }

  async doContract(transaction: DirectSignRequest): Promise<WalletResponse> {
    return this.approvedOperation('/api/controller/signing/requests', { kind: 'transaction', chain_id: transaction.chainId, transaction }, WalletResponseType.TRANSACTION);
  }

  async invokeSkill(params: SkillInvocationParams): Promise<WalletResponse> {
    const transaction: DirectSignRequest = {
      action: {
        action: 'knirv.skill.invoke',
        sender: params.sender,
        recipient: params.skill_id,
        amount: params.amount,
        payload: new TextEncoder().encode(JSON.stringify({ user_id: params.user_id, metadata: params.metadata ?? {} })),
        timestampUnix: Math.floor(Date.now() / 1000),
      },
      chainId: this.chainId,
      accountNumber: params.account_number,
      sequence: params.sequence,
      fee: params.fee,
    };
    return this.approvedOperation('/api/controller/signing/requests', { kind: 'transaction', chain_id: this.chainId, transaction }, WalletResponseType.TRANSACTION);
  }

  async resolveKnirvURI(uri: string): Promise<WalletResponse> {
    const data = await this.request(`/api/transmission/resolve?uri=${encodeURIComponent(uri)}`);
    return { code: 0, status: WalletResponseStatus.SUCCESS, type: WalletResponseType.TRANSACTION, data };
  }

  async broadcastToNetwork(data: unknown): Promise<WalletResponse> {
    const result = await this.request('/api/transmission/broadcast', { method: 'POST', body: JSON.stringify(data) });
    return { code: 0, status: WalletResponseStatus.SUCCESS, type: WalletResponseType.TRANSACTION, data: result };
  }
}

export { KNIRVWallet as AdenaWallet };

export {
  generateHDPath, Provider, TransactionEndpoint, Tx, Wallet as Tm2Wallet,
  BroadcastTxCommitResult, BroadcastTxSyncResult, TxSignature,
} from '@gnolang/tm2-js-client';
export { LedgerConnector } from '@cosmjs/ledger-amino';

File: packages/KNIRVSDK/rust/src/wallet.rs
//! Controller-custodied wallet API. User private keys never enter this process.
use crate::{
    client::HttpClient,
    error::{Error, Result},
    signing::{Action, DirectSignRequest, Fee},
    types::SkillInvocationParams,
};
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::time::{Duration, Instant};
use tokio::time::sleep;

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct WalletResponse<T = Value> {
    pub code: u16,
    pub status: WalletStatus,
    #[serde(rename = "type")]
    pub response_type: WalletResponseType,
    pub message: Option<String>,
    pub data: Option<T>,
}
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum WalletStatus {
    Success,
    Failure,
    Reject,
}
#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum WalletResponseType {
    Account,
    Network,
    Sign,
    Transaction,
}
#[derive(Deserialize)]
struct ApprovalRequest {
    request_id: String,
}
#[derive(Deserialize)]
struct ApprovalStatus {
    status: String,
    result: Option<Value>,
    reason: Option<String>,
}
#[derive(Clone, Debug)]
pub struct KnirvWallet {
    http: HttpClient,
    approval_timeout: Duration,
    poll_interval: Duration,
}
impl KnirvWallet {
    pub fn new(http: HttpClient) -> Self {
        Self {
            http,
            approval_timeout: Duration::from_secs(300),
            poll_interval: Duration::from_millis(1500),
        }
    }
    pub fn with_approval_timing(mut self, timeout: Duration, interval: Duration) -> Self {
        self.approval_timeout = timeout;
        self.poll_interval = interval;
        self
    }
    pub async fn account(&self) -> Result<WalletResponse> {
        let data = self.http.get("/api/controller/account", &[]).await?;
        Ok(success(WalletResponseType::Account, data))
    }
    pub async fn add_network(&self, network: Value) -> Result<WalletResponse> {
        self.approved(
            "/api/controller/networks",
            network,
            WalletResponseType::Network,
        )
        .await
    }
    pub async fn switch_network(&self, chain_id: &str) -> Result<WalletResponse> {
        self.approved(
            "/api/controller/networks/switch",
            json!({"chain_id":chain_id}),
            WalletResponseType::Network,
        )
        .await
    }
    pub async fn sign_transaction(
        &self,
        transaction: &DirectSignRequest,
    ) -> Result<WalletResponse> {
        self.approved(
            "/api/controller/signing/requests",
            json!({"kind":"transaction","chain_id":transaction.chain_id,"transaction":transaction}),
            WalletResponseType::Sign,
        )
        .await
    }
    pub async fn sign_message(&self, envelope: Value) -> Result<WalletResponse> {
        self.approved(
            "/api/controller/signing/requests",
            json!({"kind":"message","envelope":envelope}),
            WalletResponseType::Sign,
        )
        .await
    }
    pub async fn do_contract(&self, transaction: &DirectSignRequest) -> Result<WalletResponse> {
        self.approved(
            "/api/controller/signing/requests",
            json!({"kind":"transaction","chain_id":transaction.chain_id,"transaction":transaction}),
            WalletResponseType::Transaction,
        )
        .await
    }
    pub async fn invoke_skill(&self, params: &SkillInvocationParams) -> Result<WalletResponse> {
        let action = Action {
            schema_version: String::new(),
            action: "knirv.skill.invoke".into(),
            sender: params.sender.clone(),
            recipient: Some(params.skill_id.clone()),
            amount: params.amount.parse().ok(),
            payload: params
                .metadata
                .as_ref()
                .map(|m| serde_json::to_vec(m).unwrap_or_default()),
            timestamp_unix: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_secs())
                .unwrap_or(0),
        };
        let fee = match &params.fee {
            Some(f) => Fee {
                denom: f.denom.clone(),
                amount: f.amount.clone(),
                gas_limit: f.gas_limit,
                payer: f.payer.clone(),
                granter: f.granter.clone(),
            },
            None => Fee::default(),
        };
        let request = DirectSignRequest {
            action,
            chain_id: "knirv-1".into(),
            account_number: params.account_number,
            sequence: params.sequence,
            fee,
        };
        self.approved(
            "/api/controller/signing/requests",
            json!({"kind":"transaction","chain_id":"knirv-1","transaction":request}),
            WalletResponseType::Transaction,
        )
        .await
    }
    pub async fn resolve_uri(&self, uri: &str) -> Result<WalletResponse> {
        let data = self
            .http
            .get("/api/transmission/resolve", &[("uri", uri.into())])
            .await?;
        Ok(success(WalletResponseType::Transaction, data))
    }
    pub async fn broadcast(&self, data: Value) -> Result<WalletResponse> {
        let data = self.http.post("/api/transmission/broadcast", data).await?;
        Ok(success(WalletResponseType::Transaction, data))
    }
    async fn approved(
        &self,
        path: &str,
        body: Value,
        kind: WalletResponseType,
    ) -> Result<WalletResponse> {
        let approval: ApprovalRequest = self.http.post(path, body).await?;
        let deadline = Instant::now() + self.approval_timeout;
        while Instant::now() < deadline {
            let state: ApprovalStatus = self
                .http
                .get(
                    &format!("/api/controller/signing/requests/{}", approval.request_id),
                    &[],
                )
                .await?;
            match state.status.as_str() {
                "approved" => return Ok(success(kind, state.result.unwrap_or(Value::Null))),
                "rejected" => {
                    return Ok(WalletResponse {
                        code: 1,
                        status: WalletStatus::Reject,
                        response_type: kind,
                        message: state.reason,
                        data: None,
                    })
                }
                "expired" => {
                    return Ok(WalletResponse {
                        code: 1,
                        status: WalletStatus::Failure,
                        response_type: kind,
                        message: Some("KNIRVCONTROLLER approval expired".into()),
                        data: None,
                    })
                }
                _ => sleep(self.poll_interval).await,
            }
        }
        Err(Error::Timeout)
    }
}
fn success(kind: WalletResponseType, data: Value) -> WalletResponse {
    WalletResponse {
        code: 0,
        status: WalletStatus::Success,
        response_type: kind,
        message: None,
        data: Some(data),
    }
}
